### PR TITLE
feat(users): setup user authentication methods schema and apis

### PR DESCRIPTION
**Pull request by apoorvdixit88**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
The PR:
- Sets up schema for user authentication methods
- adds queries to interact with schema
- adds endpoints to create, update and list authentication methods

### Additional Changes

- [x] This PR modifies the API contract
- [x] This PR modifies the database schema
- [x] This PR modifies application configuration/environment variables

1. `config/config.example.toml`
2. `config/deployments/env_specific.toml`
3. `config/development.toml`
4. `config/docker_compose.toml`

## Motivation and Context
Closes [#4998](https://github.com/juspay/hyperswitch/issues/4998)

## How did you test it?
To create authentication method, ( auth is admin_api_key only):
```
curl --location 'http://localhost:8080/user/auth' \
--header 'Content-Type: application/json' \
--header 'api-key: test_admin' \
--header 'Cookie: Cookie_1=value' \
--data '
{
"owner_id": "pw",
    "owner_type": "tenant",
    "auth_method": {
        "auth_type": "open_id_connect",
        "private_config": {
            "base_url": "hello.com",
            "client_id": "something",
            "client_secret": "something",
            "private_key": "something"
        },
        "public_config": {
            "name": "okta"
        }
    },
    "allow_signup": true
    }'
```
Response: 200 OK if auth method got created successfully.

To update
```
curl --location --request PUT 'http://localhost:8080/user/auth' \
--header 'Content-Type: application/json' \
--header 'api-key: test_admin' \
--header 'Cookie: Cookie_1=value' \
--data '{
    "id": "65ff02eb-f3dc-4afc-b8ac-cef2a30c4d12",
    "auth_method": {
        "auth_type": "open_id_connect",
        "private_config": {
            "base_url": "hello.com",
            "client_id": "something",
            "client_secret": "something_update",
            "private_key": "something_update"
        },
        "public_config": {
            "name": "okta"
        }
    }
}
'
```
Response will be 200 OK for success full auth config update

To get list of authentication methods
```
curl --location 'http://localhost:8080/user/auth/list?auth_id=cbd4e678-6c09-4211-8611-74700aa1f695' \
--header 'Cookie: Cookie_1=value' \
--data ''
```
Response
```
[
    {
        "id": "6b543648-4d47-498d-bcc2-68132dbf8521",
        "auth_id": "cbd4e678-6c09-4211-8611-74700aa1f695",
        "auth_method": {
            "type": "open_id_connect",
            "name": "okta"
        },
        "allow_signup": true
    },
    {
        "id": "1364c40d-1c5a-4dc9-ae28-ce908f537e83",
        "auth_id": "cbd4e678-6c09-4211-8611-74700aa1f695",
        "auth_method": {
            "type": "password",
            "name": null
        },
        "allow_signup": true
    }
]
```

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

**Comment by racnan** [comment]
- `auth_config` should not be exposed.
- Create a `auth_config` type in application.

File: migrations/2024-06-10-084722_create_org_authentication_table/up.sql [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+CREATE TABLE IF NOT EXISTS org_authentication_methods (
+    id SERIAL PRIMARY KEY,
+	org_id VARCHAR(64) NOT NULL, 
+    auth_method "AuthMethod" NOT NULL,
+    auth_config JSONB,
+    created_at TIMESTAMP NOT NULL DEFAULT now(),
+    last_modified_at TIMESTAMP NOT NULL DEFAULT now(),
+	CONSTRAINT org_auth_method_unique UNIQUE (org_id, auth_method)
+);
```
Comment (ThisIsMani) [inline migrations/2024-06-10-084722_create_org_authentication_table/up.sql:16]: Fix spaces

File: migrations/2024-06-10-084722_create_org_authentication_table/down.sql [hunk 2]
```diff
@@ -0,0 +1,5 @@
+-- This file should undo anything in `up.sql`
```
Comment (ThisIsMani) [inline migrations/2024-06-10-084722_create_org_authentication_table/down.sql:1]: Do we need to remove `org_auth_method_unique` constraint as well? Or will it be automatically removed?

File: crates/router/src/routes/user.rs [hunk 3]
```diff
@@ -749,3 +749,60 @@ pub async fn check_two_factor_auth_status(
     ))
     .await
 }
+
+pub async fn create_org_authentication_method(
+    state: web::Data<AppState>,
+    req: HttpRequest,
+    json_payload: web::Json<user_api::CreateOrgAuthenticationMethodRequest>,
+) -> HttpResponse {
+    let flow = Flow::ListOrgAuthenticationMethods;
```
Comment (ThisIsMani) [inline crates/router/src/routes/user.rs:758]: Fix the `Flow`.

File: crates/router/src/routes/user.rs [hunk 4]
```diff
@@ -749,3 +749,60 @@ pub async fn check_two_factor_auth_status(
...
+    ))
+    .await
+}
+
+pub async fn update_org_authentication_method(
+    state: web::Data<AppState>,
+    req: HttpRequest,
+    json_payload: web::Json<user_api::UpdateOrgAuthenticationMethodRequest>,
+) -> HttpResponse {
+    let flow = Flow::ListOrgAuthenticationMethods;
```
Comment (ThisIsMani) [inline crates/router/src/routes/user.rs:777]: Fix the `Flow`.

File: migrations/2024-06-10-084722_create_org_authentication_table/up.sql [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+CREATE TABLE IF NOT EXISTS org_authentication_methods (
+    id SERIAL PRIMARY KEY,
+	org_id VARCHAR(64) NOT NULL, 
+    auth_method "AuthMethod" NOT NULL,
+    auth_config JSONB,
+    created_at TIMESTAMP NOT NULL DEFAULT now(),
+    last_modified_at TIMESTAMP NOT NULL DEFAULT now(),
+	CONSTRAINT org_auth_method_unique UNIQUE (org_id, auth_method)
```
Comment (ThisIsMani) [inline migrations/2024-06-10-084722_create_org_authentication_table/up.sql:15]: Where are we going to store info like `totp_required`, `allow_signups`?

If we are going to store them in `auth_config`, why are we using `serde::Value` in the application? I think it should be a proper struct, atleast for the above mentioned fields.

Probably like this:
```rust
struct AuthConfig {
	totp_required: bool,
	allow_signups: bool,
	metadata: Option<serde::Value>
}
```

File: crates/router/src/core/errors/user.rs [hunk 6]
```diff
@@ -80,6 +80,10 @@ pub enum UserErrors {
     TwoFactorAuthNotSetup,
     #[error("TOTP secret not found")]
     TotpSecretNotFound,
+    #[error("Org auth method already exists")]
```
Comment (ThisIsMani) [inline crates/router/src/core/errors/user.rs:83]: Please match with the error message.

File: crates/router/src/core/errors/user.rs [hunk 7]
```diff
@@ -80,6 +80,10 @@ pub enum UserErrors {
...
+    OrgAuthMethodAlreadyExists,
+    #[error("Invalid org auth method operation")]
```
Comment (ThisIsMani) [inline crates/router/src/core/errors/user.rs:85]: This as well.

File: crates/router/src/db/org_authentication_method.rs [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+    ) -> CustomResult<storage::OrgAuthenticationMethod, errors::StorageError> {
+        let conn = connection::pg_connection_write(self).await?;
+        org_authentication_method
+            .insert(&conn)
+            .await
+            .map_err(|error| report!(errors::StorageError::from(error)))
+    }
+
+    #[instrument(skip_all)]
+    async fn get_org_authentication_methods_details(
```
Comment (ThisIsMani) [inline crates/router/src/db/org_authentication_method.rs:50]:
```suggestion
    async fn list_authentication_methods_for_org_id(
```

File: crates/router/src/db/org_authentication_method.rs [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+        &self,
+        org_authentication_method: storage::OrgAuthenticationMethodNew,
+    ) -> CustomResult<storage::OrgAuthenticationMethod, errors::StorageError> {
+        let mut org_authentication_methods = self.org_authentication_methods.lock().await;
+        if org_authentication_methods
+            .iter()
+            .any(|auth_method_inner| auth_method_inner.org_id == org_authentication_method.org_id)
+        {
+            Err(errors::StorageError::DuplicateValue {
+                entity: "role_id",
```
Comment (ThisIsMani) [inline crates/router/src/db/org_authentication_method.rs:91]:
```suggestion
                entity: "org_id",
```

**Review by ThisIsMani — APPROVED** [review]
Should we encrypt `auth_configs` in DB? It might contain merchant specific secrets.

**Review by ThisIsMani — COMMENTED** [review]
Are we sure that we are going to create separate table to tenants?

File: migrations/2024-06-10-084722_create_org_authentication_table/down.sql (hunk 2 above)
Comment (apoorvdixit88) [reply migrations/2024-06-10-084722_create_org_authentication_table/down.sql:1]: dropping the org_authentication_methods table will also remove the org_auth_method_unique constraint, so we typically don't need to explicitly drop the constraint

File: crates/diesel_models/src/org_authentication_method.rs [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+use crate::{enums, schema::org_authentication_methods};
+
+#[derive(Clone, Debug, Identifiable, Queryable)]
+#[diesel(table_name = org_authentication_methods)]
+pub struct OrgAuthenticationMethod {
+    pub id: i32,
+    pub owner_id: String,
+    pub auth_method: enums::AuthMethod,
+    pub config: Option<serde_json::Value>,
```
Comment (ThisIsMani) [inline crates/diesel_models/src/org_authentication_method.rs:12]: Are we not making this a proper struct type?

File: crates/router/src/core/user.rs [hunk 11]
```diff
@@ -875,7 +876,7 @@ pub async fn resend_invite(
             if e.current_context().is_db_not_found() {
                 e.change_context(UserErrors::InvalidRoleOperation)
                     .attach_printable(format!(
-                        "User role with user_id = {} and org_id = {} is not found",
+                        "User role with user_id = {} and owner_id = {} is not found",
```
Comment (ThisIsMani) [inline crates/router/src/core/user.rs:879]: Why not
```suggestion
                        "User role with user_id = {} and merchant_id = {} is not found",
```

File: crates/router/src/core/user.rs [hunk 12]
```diff
@@ -1144,7 +1145,7 @@ pub async fn switch_merchant_id(
                 }
             })?;
 
-        let org_id = state
+        let owner_id = state
```
Comment (ThisIsMani) [inline crates/router/src/core/user.rs:1148]: Why this change is needed? `org_id` is good enough imo.

File: crates/router/src/core/user.rs [hunk 13]
```diff
@@ -1982,3 +1983,133 @@ pub async fn check_two_factor_auth_status(
...
+            .encryption_key
+            .clone()
+            .expose(),
+    )
+    .change_context(UserErrors::InternalServerError)
+    .attach_printable("Failed to decode DEK")?;
+
+    let auth_config = serde_json::to_value(req.config)
+        .change_context(UserErrors::InternalServerError)
+        .attach_printable("Failed to convert auth config to json")?;
```
Comment (ThisIsMani) [inline crates/router/src/core/user.rs:2005]: I don't think we should throw 500 here, should be a 4xx.

File: crates/router/src/core/user.rs [hunk 14]
```diff
@@ -1982,3 +1983,133 @@ pub async fn check_two_factor_auth_status(
...
+
```
Comment (ThisIsMani) [inline crates/router/src/core/user.rs:2068]: here as well

**Review by ThisIsMani — COMMENTED** [review]
Where are we going to store `totp_required`? Neither we have a column in the table nor `AuthConfig` have any field for this.

File: crates/api_models/src/user.rs [hunk 15]
```diff
@@ -280,3 +280,53 @@ pub struct VerifyRecoveryCodeRequest {
...
+
+#[derive(Debug, serde::Deserialize, serde::Serialize)]
+pub struct UserAuthenticationMethodResponse {
+    pub id: String,
+    pub auth_id: String,
+    pub owner_id: String,
+    pub owner_type: common_enums::Owner,
+    pub auth_method: common_enums::AuthMethod,
+    pub allow_signup: bool,
+}
```
Comment (ThisIsMani) [inline crates/api_models/src/user.rs:332]: Where are we sending the name of OIDC here? How will FE know that they should show okta when they get `open_id_connect` as value?

File: crates/router/src/core/user.rs [hunk 16]
```diff
@@ -1982,3 +1983,133 @@ pub async fn check_two_factor_auth_status(
...
+        .list_user_authentication_methods_for_owner_id(&req.owner_id)
+        .await
+        .change_context(UserErrors::InternalServerError)
+        .attach_printable("Failed to get list of auth methods for the owner id")?;
+
+    let auth_id = if let Some(auth_method) = auth_methods.first() {
+        auth_method.auth_id.clone()
+    } else {
+        uuid::Uuid::new_v4().to_string()
+    };
```
Comment (ThisIsMani) [inline crates/router/src/core/user.rs:2027]:
```suggestion
    let auth_id = auth_methods
	.first()
	.map(|auth_method| auth_method.auth_id.clone())
	.unwrap_or(uuid::Uuid::new_v4().to_string());

File: migrations/2024-06-10-084722_create_user_authentication_methods_table/down.sql [hunk 17]
```diff
@@ -0,0 +1,4 @@
+-- This file should undo anything in `up.sql`
+DROP INDEX IF EXISTS auth_id;
+DROP INDEX IF EXISTS owner_id;
```
Comment (ThisIsMani) [inline migrations/2024-06-10-084722_create_user_authentication_methods_table/down.sql:3]: Names doesn't match with `up.sql`. Will this work?

**Comment by racnan** [comment]
> Where are we going to store `totp_required`? Neither we have a column in the table nor `AuthConfig` have any field for this.

@ThisIsMani for now we can assume `totp_required: true` for password and magic link flow and `totp_required: false` for sso. will add a column/config later when exceptions come.

**Review by ThisIsMani — DISMISSED** [review]

**Review by SanchithHegde — APPROVED** [review]

File: crates/router/src/core/user.rs [hunk 18]
```diff
@@ -1982,3 +1984,190 @@ pub async fn check_two_factor_auth_status(
...
+        .await
+        .change_context(UserErrors::InvalidUserAuthMethodOperation)?;
+    Ok(ApplicationResponse::StatusOk)
+}
+
+pub async fn list_user_authentication_methods(
+    state: SessionState,
+    req: user_api::GetUserAuthenticationMethodsRequest,
+) -> UserResponse<Vec<user_api::UserAuthenticationMethodResponse>> {
+    let user_auth_encryption_key = hex::decode(
```
Comment (inventvenkat) [inline crates/router/src/core/user.rs:2104]: We should not do KMS decode every time, it will increase the latency and the cost. We can instead cache the decoded data. 

Let's reconsider encrypting the whole config payload : Do we really need to encrypt the whole config data? Because the response needs only non sensitive data from config. By eliminating encryption part in non sensitive data, we can skip this decryption part and also the KMS part in this API.

**Review by inventvenkat — CHANGES_REQUESTED** [review]

File: crates/api_models/src/user.rs [hunk 19]
```diff
@@ -280,3 +280,60 @@ pub struct VerifyRecoveryCodeRequest {
...
+    pub config: AuthConfig,
+}
+
+#[derive(Debug, serde::Deserialize, serde::Serialize)]
+pub struct GetUserAuthenticationMethodsRequest {
+    pub auth_id: String,
+}
+
+#[derive(Debug, serde::Deserialize, serde::Serialize)]
+pub struct UserAuthenticationMethodResponse {
```
[the remark this answers is not in the document]
Comment (apoorvdixit88) [reply crates/api_models/src/user.rs:345]: we will be returning `Vec` of the response in list, and also this singular response can used for other purposes. List can contain collection of this response type.

File: crates/router/src/core/user.rs [hunk 20]
```diff
@@ -1982,3 +1984,184 @@ pub async fn check_two_factor_auth_status(
...
+            .user_auth_methods
+            .get_inner()
+            .encryption_key
+            .clone()
+            .expose(),
+    )
+    .change_context(UserErrors::InternalServerError)
+    .attach_printable("Failed to decode DEK")?;
+
+    let (private_config, public_config) = match req.auth_method {
```
Comment (racnan) [inline crates/router/src/core/user.rs:2082]: repeated code.

**Review by inventvenkat — APPROVED** [review]

**Review by racnan — APPROVED** [review]

File: crates/diesel_models/src/user_authentication_method.rs [hunk 21]
```diff
@@ -0,0 +1,65 @@
...
+    pub auth_id: String,
+    pub owner_id: String,
+    pub owner_type: enums::Owner,
+    pub auth_type: enums::UserAuthType,
+    pub private_config: Option<Encryption>,
+    pub public_config: Option<serde_json::Value>,
+    pub allow_signup: bool,
+    pub created_at: PrimitiveDateTime,
+    pub last_modified_at: PrimitiveDateTime,
+}
```
[the remark this answers is not in the document]
Comment (apoorvdixit88) [reply crates/diesel_models/src/user_authentication_method.rs:34]: In our diesel models we have like this, we use `Insertable` for `new` type and use the other as response type. Better to keep this distinction between types and keep things consistent.

File: crates/router/src/core/user.rs (hunk 20 above)
Comment (apoorvdixit88) [reply crates/router/src/core/user.rs:2082]: we can cover this along with constraint addition in upcoming PR

File: crates/api_models/src/user.rs [hunk 22]
```diff
@@ -280,3 +280,78 @@ pub struct VerifyRecoveryCodeRequest {
...
+    pub auth_id: String,
+    pub auth_method: AuthMethodDetails,
+    pub allow_signup: bool,
+}
+
+#[derive(Debug, serde::Deserialize, serde::Serialize)]
+pub struct AuthMethodDetails {
+    #[serde(rename = "type")]
+    pub auth_type: common_enums::UserAuthType,
+    pub name: Option<OpenIdProvider>,
```
[the remark this answers is not in the document]
Comment (apoorvdixit88) [reply crates/api_models/src/user.rs:356]: This can be changed to `String` when we will add other cases, where name doesn't come from `OpenIdProvider`

File: crates/common_enums/src/enums.rs [hunk 23]
```diff
@@ -2768,3 +2768,45 @@ pub enum TokenPurpose {
...
+    strum::Display,
+    strum::EnumString,
+)]
+#[router_derive::diesel_enum(storage_type = "text")]
+#[strum(serialize_all = "snake_case")]
+#[serde(rename_all = "snake_case")]
+pub enum Owner {
+    Organization,
+    Tenant,
+    Internal,
```
[the remark this answers is not in the document]
Comment (apoorvdixit88) [reply crates/common_enums/src/enums.rs:2811]: we can remove `Internal`, or assign some id (`hyperswitch`) if we think we have relevant use cases for this.
